library/std/src/io/mod.rs

```diff
@@ -1953,6 +1953,37 @@ pub trait BufRead: Read {
     #[stable(feature = "rust1", since = "1.0.0")]
     fn consume(&mut self, amt: usize);
 
+    /// Check if the underlying `Read` has any data left to be read.
+    ///
+    /// This function may fill the buffer to check for data,
+    /// so this functions returns `Result<bool>`, not `bool`.
+    ///
+    /// Default implementation calls `fill_buf` and checks that
+    /// returned slice is empty (which means that there is no data left,
+    /// since EOF is reached).
+    ///
+    /// Examples
+    ///
+    /// ```
+    /// #![feature(buf_read_has_data_left)]
+    /// use std::io;
+    /// use std::io::prelude::*;
+    ///
+    /// let stdin = io::stdin();
+    /// let mut stdin = stdin.lock();
+    ///
+    /// while stdin.has_data_left().unwrap() {
+    ///     let mut line = String::new();
+    ///     stdin.read_line(&mut line).unwrap();
+    ///     // work with line
+    ///     println!("{:?}", line);
+    /// }
+    /// ```
+    #[unstable(feature = "buf_read_has_data_left", reason = "recently added", issue = "40745")]
+    fn has_data_left(&mut self) -> Result<bool> {
```

Comment:
`has_remaining_data` seems less... informal?

Is there a reason to check the negative case, i.e. it being not empty as opposed to the `is_empty` that can be found on many other things?

Also, what's the benefit of this method if it requires error handling? If you have to do error handling anyway you might as well do, that way you only need to do error handling once instead of doing it in the loop condition and when reading the line.

```rust
let mut line = String::new();
loop {
   match std.read_line(&mut line) {
      Ok(0) => break, // eof
      Ok(_) => {
          // do something with line
      },
      Err(e) => panic!("read error {}", e);
   }
}
```

Reply:
I just couldn't think of a good name for the positive case, but I'm open for suggestions.

As for the benefits, the main use case I have for this is for things like deserializing object directly from a file in a loop. The deserialization function will return an error if it encounters an EOF, and depending on the Serde library it's not always easy to tell whether the error is caused by EOF. It's easier to check for EOF separately in the loop condition.
```
while file.has_data_left()? {
    let obj = rmp_serde::decode::from_read(&mut file)?;
    // rest of the loop
}
```
As shown above, the additional error handling isn't really a big deal when using question mark syntax (or just unwrapping).

Reply:
Does that really solve the stated problem though? serde could still encounter an EOF in the middle of parsing an input.

Reply:
If serde encounters an EOF in the middle of deserializing an object then it's a syntax error and the input is invalid, at least for that object. If serde finishes parsing the last object in a valid input then all that's left in the reader would be EOF. In that case having EOF checking would prevent serde from attempting to deserialize another object and throwing a syntax error. Having EOF checking between deserialize calls allows the loop to end gracefully when encountering EOF in a valid input.